---
title: "The RARE-Seq pipeline"
output:
  html_document: 
    toc: true
    toc_float: true
    df_print: paged
    theme: united
    highlight: tango
    tidy: true 
    code_download: true
    tocify:
      smooth_scroll: false
      hide: true
  pdf_document: default
  word_document: default
  md_document: default
---

## 1. Introduction {#1_introduction}

This tutorial is aimed at RARE-Seq pipeline users. It is structured in the following sections:

- [1. Introduction](#1_introduction)
- [2. Instructions for Pipeline Users](#2_instructions_for_pipeline_users)
    - [2.1 Overview of the RARE-Seq Pipeline](#21_overview_of_the_rare_seq_pipeline)
        - [Stage 1: Expression analysis](#stage-1-expression-analysis)
        - [Stage 2: Genotyping Analysis](#stage-2-genotyping-analysis)
    - [2.2 The Pipeline Input](#22_the_pipeline_input)
    - [2.3 Running the pipeline](#23_the_pipeline_run)
    - [2.4 The Pipeline Output](#24_the_pipeline_output)
    
## 2. Instructions for Pipeline Users {#2_instructions_for_pipeline_users}

### 2.1 Overview of the RARE-Seq Pipeline {#21_overview_of_the_rare_seq_pipeline}

The RARE-Seq pipeline is organized into the steps described below:

0. **Demultiplexing**: Splits the parent FASTQ files into individual FASTQ files for each sample.
1. **Fastp Processing**: Performs adapter sequence trimming, quality control and filtering of the sequencing reads.
2. **Mapping**: Aligns the reads to the reference genome with STAR, using a standard 2-pass mapping strategy, and extracts uniquely mapping reads.
3. **Pre-dedupe QC**: Performs various quality checks, including estimating the DNA contamination rate, rRNA percent estimation, and other stats.
4. **Deduping**: Filters out off-target reads and collapses duplicated on-target reads into barcode families. A barcode family is defined as a set of reads that share the same UMI, start positions for them and their respective read mates, as well as the splicing signature for them and their respective mates.
5. **Expression Quantification**: Applies RSEM on the deduped data to quantify gene expression.
6. **Downsampling Analysis**: Runs a downsampling analysis to estimate the library complexity. This analysis includes the following steps:
    * downsampling of the pre-deduping **on- and off-target** reads from each sample to predefined depth levels.
    * deduping of the downsampled reads.
    * calculating a downsampling curve with the number of input reads on the x-axis and the number of on-target barcode families on the y-axis.
7. **Freq file generation**: Outputs freq files that can be used for variant calling. This step is disabled by default for whole-exome selectors, but can be enabled as described below.
8. **Fusion Calling**: Runs a separate mapping of the FASTQ files with STAR-Fusion to identify fusion candidates.


### 2.2 The Pipeline Input {#22_the_pipeline_input}

The example input data from this tutorial can be downloaded from the [RARE-seq website](https://rareseq.stanford.edu/).

The pipeline needs 3 mandatory and 1 optional input files:

1. FASTQ R1 file (required): The input FASTQ R1 file, in .fastq.gz format. This file can (and most of the times should) be a symlink:
    ```{bash, eval = T}
    ls ../HiSeq1574_dowsampled/*1.fq.gz
    ```
    
2. FASTQ R2 file (required): The input FASTQ R2 file, in .fastq.gz format. This file can (and most of the times should) be a symlink:
    ```{bash, eval = T}
    ls ../HiSeq1574_dowsampled/*2.fq.gz
    ```
    
3. Sample2barcode file (required): A tab-delimited file with the following columns:
    1. Sample name.
    2. Sample barcode.
    3. Sample type. Currently restricted to: cfRNA (or cfrna), Tumor (or tumor), Normal (or normal).
    4. The barcode scheme. Currently restricted to: Flex (or flex).
    5. The selector bed file.  
    **Please note that:**
    * **this file should not have any headers.**
    * **columns should be separated by a <u>single</u> tab character, not space, or other separators.**
    * **the combination of sample names/sample types should be unique across rows (i.e., column 1 + column 3 should be unique)**.  
    * **the name of the sample2barcode file should include the "s2b" or "sample2barcode" strings.**  
Here is an example sample2barcode file: 
    ```{r}
    s2b = read.delim("../HiSeq1574_dowsampled/sample2barcode_hiseq1574.txt", header = F)
    cat(apply(head(s2b), 1, paste, collapse = "\t"), sep = "\n")
    ```
    
4. Configuration file (optional). This is a configuration file specifying the pipeline parameters. If not provided, the default configuration output below is used:
    ```{r, engine='python', code=readLines('../RARESeq-pipeline/configs/config_docker.py')}
    ```
    To provide a custom configuration to the pipeline, you need to copy the default configuration file from the following path: `configs/config_docker.py`into the folder containing the  lane's inputs.   
    Here we illustrate several parameters that you can modify in the configuration file to adjust the behavior of the pipeline. Only modify the rest if you know what you are doing, as it can lead to the pipeline crashing:    
    * Pipeline steps to skip:
    ```{python, eval = F}
     # Steps to skip. Define as vector integres
    # e.g. to skip steps 0 and 1 define 'avoid: [0,1]'
    # (-1) Skip all
    # () Skip none (default)
    # (0) Demultiplex
    # (1) fastp
    # (2) mapping
    # (3) pre-dedupe QC (DNA contamination, molecule length, read stats)
    # (4) deduping
    # (5) expression quantification
    # (6) downsampling analysis
    # (7) freq file generation 
    # (8) fusion calling   
    
    'avoid': []
    ```
    * Extension of the selector bed file. The extension of the selector file used for filtering for on-target reads. By default a 500bp padding is added around the target selector positions. If set to '', no padding is used. If other padding is set, you need to make sure that the corresponding bed file is available in the folder containing the selectors.
    ```{python, eval = F}
    'bed_ext': 'add500bp',
    ```
    * Parameters controlling the generation of freq files. By default, freq files for samples whose selector is larger than *bed_limit* base-pairs are not generated. You can set *bed_force_freq* to "True" to force the freq file generation in such cases. Please not that doing so will significantly increase the running time of the lanes that contain samples captured with large selectors:  
    ```{python, eval = F}
    'bed_force_freq': False, 
    'bed_limit': 100000000,
    ```
    * The downsampling levels:
    ```{python, eval = F}
    #Downsampling parameters
    'downsampling_breaks' : [x for x in [1e6, 2.5e6, 5e6, 1e7, 2.5e7, 5e7, 1e8, 2.5e8, 5e8]],
    ```

All these files should be available in the same folder, e.g. `/silo5/cfRNA/v3.3/HiSeq1805/`.

### 2.3 Running the Pipeline {#23_the_pipeline_run}

Run the pipeline using this command, where each of the 4 inputs are described in the previous section:

```{bash, eval = F}
python pipeline-run.py $1 $2 $3 $4
```
### 2.4 The Pipeline Output {#24_the_pipeline_output}

The output of the pipeline should include the following files inside the root directory, e.g `../HiSeq1574_dowsampled/`:

* RSEM outputs containing expression quantification in `exp_analysis/expression/barcode-deduped/*.genes.results`, e.g:
    
    ```{bash}
    cat ../HiSeq1574_dowsampled/exp_analysis/expression/barcode-deduped/Sample_LUP1004.C1.C2_cfrna.genes.results | head
    ```

* The deduped bam files in `exp_analysis/deduped/output/barcode-deduped/Sample*/Sample*.sorted.dualindex-deduped.sorted.bam`, e.g.:

    ```{bash}
    samtools view ../HiSeq1574_dowsampled/exp_analysis/deduped/output/barcode-deduped/Sample_LUP1004.C1.C2_cfrna/Sample_LUP1004.C1.C2_cfrna.sorted.dualindex-deduped.sorted.bam | head
    ```

* The freq and indel files in `exp_analysis/deduped/output/barcode-deduped/Sample*/Sample*.sorted.dualindex-deduped.sorted.freq.paired.Q30.txt` and `exp_analysis/deduped/output/barcode-deduped/Sample*/Sample*.sorted.dualindex-deduped.sorted.indels.paired.Q30.txt`, respectively, e.g.:

    ```{bash}
    cat ../HiSeq1574_dowsampled/exp_analysis/deduped/output/barcode-deduped/Sample_LUP1004.C1.C2_cfrna/Sample_LUP1004.C1.C2_cfrna.sorted.dualindex-deduped.sorted.freq.paired.Q30.txt | head
    cat ../HiSeq1574_dowsampled/exp_analysis/deduped/output/barcode-deduped/Sample_LUP1004.C1.C2_cfrna/Sample_LUP1004.C1.C2_cfrna.sorted.dualindex-deduped.sorted.indels.paired.Q30.txt | head
    ```
